File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.io.*;
import java.util.*;

class Solution {
    
    static int len;

    
    public boolean solution(String[] phone_book) {

        boolean answer = true;
        
    
        Arrays.sort(phone_book, (o1, o2) -> o1.length() - o2.length());
        
        HashSet<String> set = new HashSet();
        
        set.add(phone_book[0]);
        
        for (int i = 1; i < phone_book.length; i++) {
            for (int j = phone_book[0].length(); j <= phone_book[i].length(); j++) {
                String inner = phone_book[i].substring(0, j);
                if (set.contains(inner)) {
                    answer = false;
                    return answer;
                }
            }
            set.add(phone_book[i]);
        }
        
        
        
        return answer;
    }
}

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_book.sort(key=lambda x : len(x))

    check = True

    hs = set()
    hs.add(phone_book[0])
    for i in range(1, len(phone_book)):
        if not check:
            break
        for j in range(len(phone_book[0]), len(phone_book[i])):
            tmp = phone_book[i][:j]
            if tmp in hs:
                check = False
                break
        hs.add(phone_book[i])

    return check